### PR TITLE
Musl

### PR DESCRIPTION
**Pull request by andriuss-xilinx**
tcc is build using normal bash script instead of bash build harness because it's mostly copy-paste from kaem file and also tcc has no tar.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -266,12 +266,30 @@ cope here.
...
+`musl` is a C standard library that is lightweight, fast, simple, free, and strives to be correct
+in the sense of standards-conformance and safety. `musl` is used by some distributions of GNU/Linux
+as their C library. Our previous Mes C library was incomplete which prevented us from building many
+newer or more complex programs.
+
+`tcc` has slight problems when building and linking `musl`, so we apply a few patches. In particular,
+we replace all weak symbols with strong symbols and will patch `tcc` in the next step to ignore duplicate
+symbols.
+
+#### Part 25 tcc 0.9.27 (musl)
```
Comment (fosslinux) [inline README.md:292]: Newline after.

File: sysa/musl-1.1.24/musl-1.1.24.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    default_src_prepare
+    rm -rf src/complex
+
+    if ! test -e /dev; then
+        mkdir /dev
+    fi
+    if ! test -e /dev/null; then
+        mknod /dev/null c 1 3
+        chmod 666 /dev/null
+    fi
```
Comment (fosslinux) [inline sysa/musl-1.1.24/musl-1.1.24.sh:11]: Hm, maybe we should have a separate "create device node" stage, I think that makes more sense than doing this here.

File: sysa/tcc-0.9.27/mes-libc-patched.kaem [hunk 3]
```diff
@@ -6,9 +6,33 @@ cd src/mes-libc
 
 # Patch
 patch -Np0 -i ../../patches/mes-libc-qsort.patch
+patch -Np0 -i ../../patches/mes-libc-crt1.patch
 
 # Recompile libc
 cd ../tcc-0.9.27
 kaem --file ../../compile-libc.kaem
 
+# Recompile tcc using patched libc
```
Comment (fosslinux) [inline sysa/tcc-0.9.27/mes-libc-patched.kaem:15]: Can we just move mes-libc-patched before tinycc-patched?

File: sysa/tcc-0.9.27/tcc-musl.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+export prefix=/after
+export libdir=${prefix}/lib/musl
+export incdir=${prefix}/include/musl
+export bindir=${prefix}/bin
+
+mkdir -p ${libdir}/tcc
+
+# Compile
+
+for TCC in tcc-0.9.26 tcc-musl; do
```
Comment (fosslinux) [inline sysa/tcc-0.9.27/tcc-musl.sh:19]: Can you add a comment here soemthing along the lines of 

> We first have to recompile using tcc-0.9.26 as tcc-0.9.27 is not selfhosting, but when linked with musl it is.

File: sysa/musl-1.1.24/patches/tcc_static.patch [hunk 5]
```diff
@@ -0,0 +1,106 @@
+diff -U3 -r src/internal/syscall.h src/internal/syscall.h
```
Comment (fosslinux) [inline sysa/musl-1.1.24/patches/tcc_static.patch:1]: Patch header:

> tinycc does not support the use of the static keyword within an array length or index.

**Review by fosslinux — CHANGES_REQUESTED** [review]
Could you please also add patch headers that (even just one sentence) explain why the patch is needed, this is very helpful for documentation. I am applying this to all other patches.

File: sysa/run.sh [hunk 6]
```diff
@@ -18,4 +18,10 @@ build flex-2.5.11
 # Part 23
 build flex-2.5.14
 
+# Part 24
+build musl-1.1.24
+
+# Part 25
+tcc-0.9.27/tcc-musl.sh
```
Comment (fosslinux) [inline sysa/run.sh:25]: Also, can we use build here?

File: sysa/tcc-0.9.27/tcc-musl.sh [hunk 7]
```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
+
+set -e
```
Comment (pder) [inline sysa/tcc-0.9.27/tcc-musl.sh:3]: Could we do set -ex here?

File: README.md (hunk 1 above)
Comment (stikonas) [reply README.md:292]: fixed

File: sysa/tcc-0.9.27/mes-libc-patched.kaem (hunk 3 above)
Comment (stikonas) [reply sysa/tcc-0.9.27/mes-libc-patched.kaem:15]: @pder promised to fix it since it's not my commit.

File: sysa/tcc-0.9.27/tcc-musl.sh (hunk 4 above)
Comment (stikonas) [reply sysa/tcc-0.9.27/tcc-musl.sh:19]: done

File: sysa/musl-1.1.24/patches/tcc_static.patch (hunk 5 above)
Comment (stikonas) [reply sysa/musl-1.1.24/patches/tcc_static.patch:1]: done

File: sysa/tcc-0.9.27/tcc-musl.sh (hunk 7 above)
Comment (stikonas) [reply sysa/tcc-0.9.27/tcc-musl.sh:3]: This script will be replaced with sourced script. So not needed

File: sysa/musl-1.1.24/musl-1.1.24.sh (hunk 2 above)
Comment (stikonas) [reply sysa/musl-1.1.24/musl-1.1.24.sh:11]: musl's `./configure` seems to corrupt `/dev/null` node for some reason, so let's just create /dev here and populate it at some point later, maybe after `musl`.

Comment (fosslinux) [reply sysa/musl-1.1.24/musl-1.1.24.sh]: Can you add a comment to that effect please? `# configure script needs these nodes but is not standard/corrupts them`

Comment (fosslinux) [reply sysa/musl-1.1.24/musl-1.1.24.sh]: Wait nvm...

File: sysa/tcc-0.9.27/tcc-musl.sh [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            -D CONFIG_TCC_LIBPATHS=\"${libdir}:${libdir}/tcc\" \
+            -D CONFIG_TCC_SYSINCLUDEPATHS=\"${incdir}\" \
+            -D TCC_LIBGCC=\"${libdir}/libc.a\" \
+            -D CONFIG_TCC_STATIC=1 \
+            -D CONFIG_USE_LIBGCC=1 \
+            -D TCC_VERSION=\"0.9.27\" \
+            -D ONE_SOURCE=1 \
+            tcc.c
+
+	mkdir -p ${libdir}/tcc
```
Comment (fosslinux) [inline sysa/tcc-0.9.27/tcc-musl.sh:37]: This can just go outside the loop.

File: sysa/tcc-0.9.27/tcc-musl.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        # libtcc1.a
+        ${TCC} -c -D HAVE_CONFIG_H=1 lib/libtcc1.c
+        ${TCC} -ar cr ${libdir}/tcc/libtcc1.a libtcc1.o
+    done
+}
+
+src_install() {
+    mv ${bindir}/tcc ${bindir}/tcc-mes
+    rm ${bindir}/boot*-tcc ${bindir}/tcc-0.9.26 ${bindir}/mes-tcc
```
Comment (fosslinux) [inline sysa/tcc-0.9.27/tcc-musl.sh:46]: Can you comment here something like remove old tcc binaries.

**Review by fosslinux — CHANGES_REQUESTED** [review]
Thanks!

**Pull request merged**